vintagestory: modernize

dtomvan opened the pull request:
## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

dtomvan committed:
vintagestory: factor out `makeWrapperArgs`

Co-authored-by: ShamrockLee <shamrocklee@posteo.net>

dtomvan committed:
vintagestory: remove scripts that should not get invoked on NixOS

Not entirely sure wether this is a good idea or not.

dtomvan commented:
Wow I completely forgot about this one 😅  Hmm... I'm debating whether or not it's a good idea to update to 1.21.0-rc.2 for now, because it's most likely not as buggy as the first dotnet 8 based version last month, and, well, we need to get this darn game off of dotnet 7.

Artturin commented on pkgs/by-name/vi/vintagestory/package.nix:
```diff
@@ -75,22 +75,27 @@ stdenv.mkDerivation (finalAttrs: {
+      wrapperFlags = lib.trim ''
+        --prefix LD_LIBRARY_PATH : "${finalAttrs.runtimeLibs}" \
+        --set-default mesa_glthread true
+      '';
```
Why? this isn't modernization or an improvement.

dtomvan replied:
I figured I just did as much as possible in one fell swoop. This was duplicate, so I factored into a variable...

ShamrockLee replied:
1. `runtimeLibs` is not a Bash array without `__structuredAttrs = true`. If you reference it as a Bash array, Bash will coerce it into an array and place the original `"$runtimeLibs"` value into the first argument. Even though the result happens to be the same as referencing as "$runtimeLibs", it's semantically incorrect.

   `"$runtimeLibs"` is in fact semantically correct in the context of `__structuredAttrs = false`. If you want to make it `__structuredAttrs = true`-ready, use `"''${runtimeLibs[*]}"`.

2. What you call `wrapperFlags` is more commonly named `makeWrapperArgs`.   By using `__structuredAttrs = true`, you can pass `makeWrapperArgs` to `stdenv.mkDerivation` directly, and reference it as a Bash associative array:
   ```nix
   makeWrapperArgs = [
     "--set-default"
     "mesa_glthread"
     "true"
   ];

   preFixup = ''
     makeWrapperArgs+=(--prefix LD_LIBRARY_PATH : ''${runtimeLibs[*]})

     makeWrapper ${lib.getExe dotnet-runtime_8} $out/bin/vintagestory "''${makeWrapperArgs[@]}" \
       --add-flags $out/share/vintagestory/Vintagestory.dll

     makeWrapper ${lib.getExe dotnet-runtime_8} $out/bin/vintagestory-server "''${makeWrapperArgs[@]}" \
       --add-flags $out/share/vintagestory/VintagestoryServer.dll
   '';
   ```

JamieMagee commented:
> Wow I completely forgot about this one 😅 Hmm... I'm debating whether or not it's a good idea to update to 1.21.0-rc.2 for now, because it's most likely not as buggy as the first dotnet 8 based version last month, and, well, we need to get this darn game off of dotnet 7.

Sorry, I didn't see this before I created #436539. We're now up to 1.21.0-rc.7. I think at this point, it should be pretty stable on .NET 8

dtomvan commented:
It's okay. We can merge an update before this PR. I'm fine with the conflicts that will arise, done it before with `1.20.1` IIRC. This PR isn't that important after all.

dtomvan commented:
After said PR I'll do the conflict resolution...

dtomvan commented:
finally done so.

dtomvan commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 420002`
Commit: `b4ef6a3d269ac212318ff238568299214143f223`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>vintagestory</li>
  </ul>
</details>

dtomvan committed:
vintagestory: remove gtk2 (#410814)

Doesn't actually need it

dtomvan committed:
vintagestory: remove unused deps

Used ltrace and LD_DEBUG=libs to deduce which libraries weren't loaded
from the nix store and thus already vendored. This saves ~400 MiB in
closure size.

ShamrockLee commented on pkgs/by-name/vi/vintagestory/package.nix:
```diff
@@ -33,15 +28,10 @@ stdenv.mkDerivation rec {
   ];
 
-  runtimeLibs = lib.makeLibraryPath (
+  env.runtimeLibs = lib.makeLibraryPath (
```
Why?

```suggestion
  runtimeLibs = lib.makeLibraryPath (
```

ShamrockLee reviewed (dismissed):
I want to know why you want to pass `runtimeLibs` as an environment variable explicitly.

dtomvan committed:
vintagestory: enable versionCheckHook

Co-authored-by: ShamrockLee <shamrocklee@posteo.net>

ShamrockLee commented on pkgs/by-name/vi/vintagestory/package.nix:
```diff
@@ -116,4 +119,4 @@ stdenv.mkDerivation rec {
     mainProgram = "vintagestory";
   };
-}
+})
```
I noticed that the meta.platforms is specified as lib.platforms.linux, but it packages the x86_64-linux binary.

Should it be specified as "x86_64" instead?

dtomvan committed:
vintagestory: set `meta.platforms = ["x86_64-linux"]`

ShamrockLee commented on pkgs/by-name/vi/vintagestory/package.nix:
```diff
@@ -93,39 +86,48 @@ stdenv.mkDerivation (finalAttrs: {
+  ];
+
+  preFixup = ''
+     makeWrapperArgs+=(--prefix LD_LIBRARY_PATH : "$(IFS=:; echo ''${runtimeLibs[@]})")
```
```suggestion
     makeWrapperArgs+=(--prefix LD_LIBRARY_PATH : "$(concatStringsSep : runtimeLibs)
```

ShamrockLee reviewed (dismissed):
The `$(IFS=:; echo)` trick doesn't work. Use the `concatStringsSep` provided by `$stdenv/setup` instead.

ShamrockLee commented:
I tested with the following Nix expression to avoid the prolonged rebuild of the .NET framework on the `master` branch:

```nix
(builtins.getFlake "github:NixOS/nixpkgs/nixos-unstable").legacyPackages.${builtins.currentSystem}.callPackage
  (builtins.getFlake "github:NixOS/nixpkgs/refs/pull/420002/merge").legacyPackages.${builtins.currentSystem}.vintagestory.override
  { }
```

I saw the login screen after it finish loading the shades and heard the background music and keyboard sound effect. The Fcitx5 input method framework doesn't work, though. (This is not a regression, as it also happens on the `nixos-unstable` branch.)

I'm testing on my x86_64-linux NixOS laptop and the GNOME desktop environment.

ShamrockLee commented:
The input method issue may result from lacking certain environment variables in the wrapper. I won't block this PR for it.

ShamrockLee commented on pkgs/by-name/vi/vintagestory/package.nix:
(Not a blocker.)

Passing `runtimeLibs` as a list via `passthru` and join it with another attribute (`runtimeLibraryPath`) enhances (preserves) its overridability while ensuring the use of the library outputs.

dtomvan commented:
@shamrocklee sorry, I haven't looked back at this PR for a while now. I think this is ready now. I addressed your most recent review.
